### ISSUE TITLE
Potential Memory Leak in p_kafka_set_topic after local queue full errors

### ISSUE THREAD

**Issue by pothier-peter** [pmacct/pmacct#507]
**Description**
Hello Paolo,

I believe there's a memory leak that occurs after experiencing local queue full Kafka producer errors.

To recreate the error (quickly) I lowered the Kafka queue.buffering.max.kbytes and modified the code in bmp_process_msg_peer_up() to continuously produce messages to Kafka

+          for (;;) {
+              bmp_log_msg(peer, &bdata, &blstats, bgp_peer_log_seq_get(&bms->log_seq), event_type, config.nfacctd_bmp_msglog_output, BMP_LOG_TYPE_STATS);
+          }

Running valgrind produced this stack

   ==30082== 14,883,906 (13,965,152 direct, 918,754 indirect) bytes in 45,938 blocks are definitely lost in loss record 81 of 81
   ==30082==    at 0x4C33B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   ==30082==    by 0x21B33F: rd_calloc (rd.h:94)
   ==30082==    by 0x21B33F: rd_kafka_topic_conf_new (rdkafka_conf.c:1896)
   ==30082==    by 0x190975: p_kafka_set_topic (kafka_common.c:87)

I suspect the leak is within this code snippet

   void p_kafka_set_topic(struct p_kafka_host *kafka_host, char *topic)
   {
     kafka_host->topic_cfg = rd_kafka_topic_conf_new();
     ….
     if (kafka_host->rk && kafka_host->topic_cfg) {
      kafka_host->topic = rd_kafka_topic_new(kafka_host->rk, topic, kafka_host->topic_cfg);
      kafka_host->topic_cfg = NULL; /* rd_kafka_topic_new() destroys conf as per rdkafka.h */
     }
   }

When the local queue full error occurs, the
    p_kafka_produce_data_to_part()
function invokes
    p_kafka_close(kafka_host, TRUE);
which sets
    kafka_host->rk = NULL;
which prevents the invocation of rd_kafka_topic_new() above (and the indirect destruction of kafka_host->topic_cfg)
in subsequent calls to p_kafka_set_topic().

I made the following change in my local workspace

        if (kafka_host->rk && kafka_host->topic_cfg) {
          kafka_host->topic = rd_kafka_topic_new(kafka_host->rk, topic, kafka_host->topic_cfg);
          kafka_host->topic_cfg = NULL; /* rd_kafka_topic_new() destroys conf as per rdkafka.h */
   +    } else {
   +      rd_kafka_topic_conf_destroy(kafka_host->topic_cfg);
   +      kafka_host->topic_cfg = NULL;
        }

and although it fixed the leak, I am unsure of unattended consequences and whether it recovers due to my
somewhat artificial test method.

Your thoughts and help are greatly appreciated.

Thank you!
Peter P

**Version**
1.7.4

**Comment by paololucente** [comment]
Hi Peter ( @pothier-peter ),

Always a pleasure to read from you, always good points. Allow me a bit of time to review & try to reproduce the scenario and come back to you. You are definitely after something, i can already say that. Maybe we should fix differently though but i'd need do reproduce the scenario first. Keep you posted (hopefully shortly)!

Paolo

**Comment by paololucente** [comment]
Hi Peter ( @pothier-peter ),

It appears this has already been fixed in 1.7.4p1. So also 1.7.5, 1.7.6 and master code have this leak resolved. I came up to this commit using git blame: https://github.com/pmacct/pmacct/commit/c99c4729c6f9d19981d8e8d7116834c503eb5c5b

Paolo

**Comment by pothier-peter** [comment]
Great! Thank You Paolo.
I'll merge this commit into our local repo.

Peter P

**Comment by pothier-peter** [comment]
Hello Paolo,

The changes in p_kafka_close() from commit c99c472 are already in our local repo.

The memory leak we've seen appears to stem from the result of invoking p_kafka_set_topic() while waiting for the last_fail retry interval to expire (connection is closed, kafka_host->rk is NULL, and connection has not yet been remade). Under this condition p_kafka_close() is not called after each invocation of p_kafka_set_topic(), and thus the kafka_host->topic_cfg is leaked.

In my local workspace, I did not invoke p_kafka_close() after calling p_kafka_set_topic() under a failed condition to fix the leak, but instead destroyed kafka_host->topic_cfg in p_kafka_set_topic() if kafka_host->rk was NULL. I suppose there are multiple ways to fix this leak, perhaps at a higher level in the code (e.g. avoid calling p_kafka_set_topic while awaiting the last_fail retry interval to complete and the connection remade), and I do not feel I am in the best situation to determine that.

Thus, I do not believe it is fixed by the referenced commit c99c472.

Thanks again,
Peter P

**Comment by pothier-peter** [comment]
Hello Paolo,

Trying to think about this from a higher level, would it make sense to avoid processing BMP messages if the Kafka Producer is not connected, in order to avoid not only the memory leak, but losing BMP messages? Something like this?


         if (last_fail && ((last_fail + P_broker_timers_get_retry_interval(&bmp_daemon_msglog_kafka_host.btimers)) <= bmp_misc_db->log_tstamp.tv_sec))
           bmp_daemon_msglog_init_kafka_host();
       }
+
+      /* Ensure connection to Kafka Producer */
+      if (bmp_daemon_msglog_kafka_host->rk == NULL) {
+        goto select_again;
+      }
 #endif


Peter P

**Comment by paololucente** [comment]
Hi Peter,

Can i ask you a favor: can you test this patch in your environment to see it working?

```
diff --git a/src/kafka_common.c b/src/kafka_common.c
index abc4054b..e150f8e2 100644
--- a/src/kafka_common.c
+++ b/src/kafka_common.c
@@ -83,7 +83,7 @@ void p_kafka_unset_topic(struct p_kafka_host *kafka_host)
 
 void p_kafka_set_topic(struct p_kafka_host *kafka_host, char *topic)
 {
-  if (kafka_host) {
+  if (kafka_host && kafka_host->rk) {
     kafka_host->topic_cfg = rd_kafka_topic_conf_new();
     p_kafka_apply_topic_config(kafka_host);
 
@@ -106,7 +106,7 @@ void p_kafka_set_topic(struct p_kafka_host *kafka_host, char *topic)
     /* destroy current allocation before making a new one */
     if (kafka_host->topic) p_kafka_unset_topic(kafka_host);
 
-    if (kafka_host->rk && kafka_host->topic_cfg) {
+    if (kafka_host->topic_cfg) {
       kafka_host->topic = rd_kafka_topic_new(kafka_host->rk, topic, kafka_host->topic_cfg);
       kafka_host->topic_cfg = NULL; /* rd_kafka_topic_new() destroys conf as per rdkafka.h */
     }
```

It does make sense for me to anticipate the `kafka_host->rk` check to earlier in the function and hence prevent needles calls to `rd_kafka_topic_conf_new()` (as now only the call to `rd_kafka_topic_new()` is avoided).

Paolo

**Comment by pothier-peter** [comment]
Hello Paolo,

Looks good with this patch. Here you can see the PMBMPD daemon, after receiving the Local Queue Full

```
2021-06-08T12:16:03Z ERROR ( 198.19.19.151-1/core ): Failed to produce to topic pmacct.bmp-msglog partition -1: Local: Queue full
2021-06-08T12:16:03Z ERROR ( 198.19.19.151-1/core ): Connection failed to Kafka: p_kafka_close()
```

runs at 100% and now memory remains low

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                      
12263 akamai    20   0  632012   9508   5300 R  99.3  0.1   0:49.82 pmbmpd1                                                      
```

Thank you,
Peter P

**Comment by paololucente** [comment]
Hi Peter,

Thanks very much for confirming: that patch is good to go then, it just makes sense & i will commit shortly later today. That 100% CPU means we can look at the second part of what you were proposing without the risk of hiding the leak; could you determine that the snippet you posted a couple of message ago (below for your reference) does help with keeping the CPU usage under control? 

```
+      /* Ensure connection to Kafka Producer */
+      if (bmp_daemon_msglog_kafka_host->rk == NULL) {
+        goto select_again;
+      } 
```

It does make sense to me, i'd only like to review a bit what does entail going back to the select there (which, i agree, it's probably anyway the most sensible thing to do); i was thinking: guess there is no BMP messages coming in or they are intermittent, we may have to wait long time before trying connecting to Kafka again (which, in turn, it may be fine as long as we retry as preps for processing the next BMP message that comes in).

Paolo

**Comment by pothier-peter** [comment]
Hello Paolo,

The 100% CPU is an artifact of the way I was testing outside of production. I hacked the code (see my original comments) to send the same Peer Up message over and over again to force the memory leak. The high CPU plus low MEM informed me that PMBMPD was still attempting to send messages through Kafka, but the leak was gone!

The "Ensure connection to Kafka Producer" snippet, I do not believe will have any effect on CPU utilization, nor on the reconnection time since it was added after the last_fail check and possible call to bmp_daemon_msglog_init_kafka_host(). Its intent was simply to avoid losing log messages through Kafka during the retry interval since I do not see where the bmp_log messages are retried on failures to Kafka. Looking back at where I placed that code, I think I should have placed it within the "if (config.nfacctd_bmp_msglog_kafka_topic) {" clause.

As for waiting a long time before reconnecting to Kafka, we have control over that via the bmp_daemon_msglog_kafka_retry configuration.

Peter P
